### PR TITLE
[wip] feat(api): bridge limit script

### PR DESCRIPTION
**Pull request by bigboydiamonds**
Description

PR introduces a new `/bridgeLimits` endpoint that provides the minimum and maximum origin input values for specific bridge routes. The endpoint allows users to query these limits by passing the following parameters:

	•	fromChain: The originating chain ID
	•	toChain: The destination chain ID
	•	fromToken: The address of the token on the originating chain
	•	toToken: The address of the token on the destination chain

Adds new `/bridgeLimits` Endpoint:
	•	The endpoint determines the min/max input values for a given bridge route.
	•	The input parameters (fromChain, toChain, fromToken, toToken) are used to match and return the appropriate route limits.
	•	returns { minOriginValue, maxOriginValue }

Adds Bridge Route Mapping:
	•	Constructs an internal mapping of all available bridge routes using following data structure:
```
{
  "originChainId": {
    "originTokenAddress": {
      "swappableType": "string",
      "symbol": "string",
      "routes": {
        "destinationChainId": {
          "destinationTokenAddress": {
            "symbol": "string",
            "minValue": "string",
            "maxValue": "string"
          }
        }
      }
    }
  }
}
```

Adds generating bridge limits script `generateLimits.ts`:
	•	script generates the min/max origin values for all supported bridge routes.
	•	script ingests JSON object at scripts/data/bridgeLimitMap.json to programatically fetch bridge quote min/max limits 

## Summary by CodeRabbit

- **New Features**
	- Introduced a new API endpoint for retrieving bridge limits between different blockchain networks and tokens.
	- Added a new script for generating limit values for token bridges based on predefined criteria.
	- Included a utility for fetching bridge quotes, enhancing the API's capabilities.
	- Added a new dependency for improved Ethereum address handling.
	- Introduced functionality for constructing a JSON representation of swappable tokens across different blockchain networks.

- **Bug Fixes**
	- Enhanced validation checks for query parameters in the new endpoints to ensure correct data types and completeness.

714ad0d92a23e6a0eae9a4320227d154f31091f8: [synapse-interface preview link ](https://sanguine-synapse-interface-q6a6wesv7-synapsecns.vercel.app)
eaa421f4c89327b00618885f50e14203235b71a3: [synapse-interface preview link ](https://sanguine-synapse-interface-4qb4y5r9s-synapsecns.vercel.app)
81d7c4791353cd4fe7833188b0d19e6a34177e59: [synapse-interface preview link ](https://sanguine-synapse-interface-8zf95ikjr-synapsecns.vercel.app)
a2095b1a3c163e9d140a18e79c724571c572f57b: [synapse-interface preview link ](https://sanguine-synapse-interface-p6rmjzoob-synapsecns.vercel.app)
39bf444d50b39720103d656be95638b01dd8d452: [synapse-interface preview link ](https://sanguine-synapse-interface-75cdkj67a-synapsecns.vercel.app)
1a8147ac850c7203f0b03c69fc9a6770e3b76549: [synapse-interface preview link ](https://sanguine-synapse-interface-mumewhqlh-synapsecns.vercel.app)
7fe5029d2521d25fded47282d6cc29395e4d42a1: [synapse-interface preview link ](https://sanguine-synapse-interface-jtq8y8arz-synapsecns.vercel.app)
fda5d62eb3a14ce37248caff9470cf1bd574cd71: [synapse-interface preview link ](https://sanguine-synapse-interface-duh6ci6go-synapsecns.vercel.app)
1be1b45c4a6ee5aedb8ba5f348abf0f56934baf4: [synapse-interface preview link ](https://sanguine-synapse-interface-a5ssl37rs-synapsecns.vercel.app)

### CODE REVIEW

**Comment by Defi-Moses** [comment]
Logic looks great to me, no structural changes that I see. The two things I think that need to be updated:
1. Validation for isTokenAddress in addition to isAddress --> gives more clarity to the developer when reading errors
2. Certain routes have weird responses regarding their decimals. Check out this example: http://localhost:3000/getBridgeLimits?fromChain=1&fromToken=0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48&toChain=81457&toToken=0x4300000000000000000000000000000000000003

Other than that seems good to me

**Comment by bigboydiamonds** [comment]
> Logic looks great to me, no structural changes that I see. The two things I think that need to be updated:
> 
> 1. Validation for isTokenAddress in addition to isAddress --> gives more clarity to the developer when reading errors
> 2. Certain routes have weird responses regarding their decimals. Check out this example: http://localhost:3000/getBridgeLimits?fromChain=1&fromToken=0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48&toChain=81457&toToken=0x4300000000000000000000000000000000000003
> 
> (the response isnt formatted for decimals properly and thus returns the incorrect maxAmountOut. My hunch is that you are parsing using fromToken decimals for the output, not toTokenDecimals.
> 
> Other than that seems good to me

Thanks for catching this! Updated in https://github.com/synapsecns/sanguine/commit/bb9aa820d56156e6e3fa95796786eed2fd2c037a , was using the `bridgeQuote.originQuery.tokenOut` decimals when it should've been the `bridgeQuote.destQuery.tokenOut` decimals to parse for max origin values

File: packages/rest-api/src/controllers/getBridgeLimitsController.ts [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import { validationResult } from 'express-validator'
+import { BigNumber } from 'ethers'
+import { parseUnits } from '@ethersproject/units'
+
+import { Synapse } from '../services/synapseService'
+import { tokenAddressToToken } from '../utils/tokenAddressToToken'
+import { formatBNToString } from '../utils/formatBNToString'
+import { BRIDGE_LIMIT_MAPPING } from '../utils/bridgeLimitMapping'
+
+export const getBridgeLimitsController = async (req, res) => {
```
Comment (abtestingalpha) [inline packages/rest-api/src/controllers/getBridgeLimitsController.ts:10]: rename to `bridgeLimitsController`

File: packages/rest-api/src/tests/getBridgeLimits.test.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import request from 'supertest'
+import express from 'express'
+
```
Comment (abtestingalpha) [inline packages/rest-api/src/tests/getBridgeLimits.test.ts:3]: rename this test file to `bridgeLimitsRoute.test.ts` to match the rest of the tests. we might write controller/middleware tests later, so this naming convention keeps them unique.

File: packages/rest-api/src/routes/bridgeLimitsRoute.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
+import express from 'express'
+import { check } from 'express-validator'
+import { isAddress } from '@ethersproject/address'
+
+import { CHAINS_ARRAY } from '../constants/chains'
+import { showFirstValidationError } from '../middleware/showFirstValidationError'
+import { getBridgeLimitsController } from '../controllers/getBridgeLimitsController'
+
+const router = express.Router()
+
```
Comment (abtestingalpha) [inline packages/rest-api/src/routes/bridgeLimitsRoute.ts:10]: add swagger docs (see other routes for examples)

File: packages/rest-api/src/routes/bridgeLimitsRoute.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+router.get(
+  '/',
+  [
+    check('fromChain')
+      .isNumeric()
```
Comment (abtestingalpha) [inline packages/rest-api/src/routes/bridgeLimitsRoute.ts:15]: existence checks should come first:

```
exists ->
is it numeric ->
is it supported
```

File: packages/rest-api/src/tests/getBridgeLimits.test.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+import getBridgeLimitsRoute from '../routes/bridgeLimitsRoute'
+
+const app = express()
+app.use('/getBridgeLimits', getBridgeLimitsRoute)
+
+describe('Get Bridge Limits Route', () => {
+  it('should return min/max origin amounts for valid input', async () => {
+    const response = await request(app).get('/getBridgeLimits').query({
+      fromChain: 1,
+      fromToken: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
```
Comment (abtestingalpha) [inline packages/rest-api/src/tests/getBridgeLimits.test.ts:13]: import `USDC` from bridgeable vs. hardcoding address where we already have it in app

File: packages/rest-api/src/tests/getBridgeLimits.test.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      toChain: 10,
+      toToken: '0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85',
+    })
+
+    expect(response.status).toBe(200)
+    expect(response.body).toHaveProperty('maxOriginAmount')
+    expect(response.body).toHaveProperty('minOriginAmount')
+  }, 10_000)
+
```
Comment (abtestingalpha) [inline packages/rest-api/src/tests/getBridgeLimits.test.ts:22]: Add tests not just for stables but also ETH.

**Pull request closed without merging**